Dockerfile

```diff
@@ -1,25 +1,78 @@
 FROM python:3.6-slim
-COPY setup.* /opt/airflow/
-COPY airflow /opt/airflow/airflow
 
-ARG AIRFLOW_HOME=/usr/local/airflow
-ARG AIRFLOW_DEPS="all"
-ARG PYTHON_DEPS=""
-ARG buildDeps="freetds-dev libkrb5-dev libsasl2-dev libssl-dev libffi-dev libpq-dev git"
-ARG APT_DEPS="$buildDeps libsasl2-dev freetds-bin build-essential default-libmysqlclient-dev apt-utils curl rsync netcat locales"
-
-RUN set -x \
-    && export SLUGIFY_USES_TEXT_UNIDECODE=yes \
-    && apt update \
-    && if [ -n "${APT_DEPS}" ]; then apt install -y $APT_DEPS; fi \
-    && if [ -n "${PYTHON_DEPS}" ]; then pip install ${PYTHON_DEPS}; fi \
-    && cd /opt/airflow \
-    && pip install -e .[$AIRFLOW_DEPS] \
-    && apt-get purge --auto-remove -yqq $buildDeps \
-    && apt-get autoremove -yqq --purge \
+SHELL ["/bin/bash", "-c"]
+
+# Make sure noninteractie debian install is used
+ENV DEBIAN_FRONTEND=noninteractive
+
+# Increase the value to force renstalling of all apt-get dependencies
+ENV FORCE_REINSTALL_APT_GET_DEPENDENCIES=1
+
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    libkrb5-dev libsasl2-dev libssl-dev libffi-dev libpq-dev git \
+    && apt-get clean
+
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    libsasl2-dev freetds-bin build-essential default-libmysqlclient-dev apt-utils curl rsync netcat locales \
     && apt-get clean
 
+# Increase the value here to force upgrade of all apt-get dependencies
+ENV FORCE_UPGRADE_OF_APT_GET=1
+
+RUN apt-get update \
+    && apt-get upgrade -y --no-install-recommends \
+    && apt-get clean
+
+# You need this in order to avoid GPL in your dependenciesd
+ARG AIRFLOW_HOME=/usr/local/airflow
 RUN mkdir -p $AIRFLOW_HOME
+
+# Airflow extras to be installed
+ARG AIRFLOW_EXTRAS="all"
+
+# Increase the value here to force reinstalling Apache Airflow pip dependencies
+ENV FORCE_REINSTALL_ALL_PIP_DEPENDENCIES=1
+
+# Speeds up building the image - cassandra driver without CYTHON saves around 10 minutes of build on typical machine
+ARG CASS_DRIVER_NO_CYTHON_ARG=1
+
+# Build cassandra driver on multiple CPUs
+ENV CASS_DRIVER_BUILD_CONCURRENCY=8
+
+# Speeds up the installation of cassandra driver
+ENV CASS_DRIVER_NO_CYTHON=${CASS_DRIVER_NO_CYTHON_ARG}
+
+## Airflow requires this variable be set on installation to avoid a GPL dependency.
+ENV SLUGIFY_USES_TEXT_UNIDECODE yes
+
+# Airflow sources change frequently but dependency onfiguration won't change that often
+# We copy setup.py and other files needed to perform setup of dependencies
+# This way cache here will only be invalidated if any of the
+# version/setup configuration change but not when airflow sources change
+ADD setup.* /opt/airflow/
+ADD airflow/version.py /opt/airflow/airflow/version.py
+ADD airflow/__init__.py /opt/airflow/airflow/__init__.py
+ADD airflow/bin/airflow /opt/airflow/airflow/bin/airflow
```

Comment:
This will create the 4th layer.  But command `ADD` supports many sources, so you can do it in one step.
```
ADD setup.* airflow/version.py airflow/__init__.py airflow/bin/airflow /opt/airflow/
````
I'm also wondering why did you use ADD instead of COPY?

Reference: 
https://docs.docker.com/engine/reference/builder/#add

Reply:
I actually changed it to COPY in the  before you asked, following best practices (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy) but I have not pushed it yet. I did it just now.

Also from the same best practices:

"If you have multiple Dockerfile steps that use different files from your context, COPY them individually, rather than all at once. This ensures that each step’s build cache is only invalidated (forcing the step to be re-run) if the specifically required files change."